### PR TITLE
[Feature] DeepSeek-V4 for Ascend950

### PR DESCRIPTION
**Pull request by nomewang**
###  What this PR does / why we need it?
  
  This PR introduces DeepSeek-V4 model support for the Ascend A5 (Ascend950) device. It adds several key capabilities:

  1. W4A8 MXFP (Microscaling) Quantization
  - Introduces a new QuantType.W4A8MXFP quantization type with dedicated linear and MoE methods (vllm_ascend/quantization/methods/w4a8_mxfp4.py,
  vllm_ascend/quantization/methods/fp8.py).
  - Adds AscendFp8Config (vllm_ascend/quantization/fp8_config.py) to serve as the quantization configuration entry point for DeepSeek-V4 FP8 models on
  Ascend, integrating with both the "fp8" and "deepseek_v4_fp8" quantization method names.
  - Implements A5-specific MXFP quantized matmul and grouped matmul-swish-glu operators in the device adaptor layer (vllm_ascend/device/device_op.py),
  leveraging NPU hardware instructions for float4_e2m1fn_x2 weight formats with float8_e8m0fnu microscaling.
  
  2. Ascend A5 Attention Kernel Optimizations
  - Major overhaul of the direct sparse attention (DSA) v1 kernel (vllm_ascend/attention/dsa_v1.py, +454/-188) to support A5-specific block sizes, scheduling
  metadata, and KV cache interfaces.
  - A5-specific compile-time/scheduling adaptations for decode and prefill attention paths.
  
  3. DeepSeek-V4 Model Architecture Support
  - Extends DeepseekV4MoE and related modules with swiglu_limit support and A5-specific quant_config handling for compressor projections
  (vllm_ascend/models/deepseek_v4.py).
  - Adjusted compressor block_size from 32 to 16 for A5 device compatibility.
  
  4. FusedMoE Enhancements
  - Updated the FusedMoE pipeline (contracts, stage params, prepare/finalize) to pass through W4A8 MXFP quantization parameters (activation quantization
  dtype, weight quantization dtype, scale dtypes, BF16 usage flag).
  
###  Why we need it: DeepSeek-V4 is a new generation model that cannot run on Ascend A5 without these adaptations. The A5 NPU has specific hardware requirements
  for weight layout, quantization format, and kernel block sizes that differ from previous Ascend generations. This PR is the first step to enable
  DeepSeek-V4 inference on Ascend950 hardware.

  ---
  Does this PR introduce any user-facing change?
  
  Yes:
  - DeepSeek-V4 can now be served on Ascend A5 hardware with W4A8 MXFP quantization by specifying --quantization fp8 (or deepseek_v4_fp8) in the vLLM serve
  command.
  - A new quantization type W4A8MXFP is available in the Ascend quantization registry.
  - No API or CLI changes; the feature is activated automatically when the Ascend device type is A5 and the model config specifies a supported quantization
  format.

  ---
 ### How was this patch tested?
  
  - Unit tests: updated tests/ut/ops/test_moe_runtime_args.py to cover the new W4A8 MXFP quantization path in MoE runtime argument construction.
  - The PR is labeled module:tests, module:ops, module:core, and module:quantization, indicating test coverage across all modified subsystems.
  - End-to-end validation was performed by running DeepSeek-V4 inference on Ascend A5 hardware with W4A8 MXFP quantized weights.

  ---

- vLLM version: v0.20.2
- vLLM main: https://github.com/vllm-project/vllm/commit/9090368b650896bf5fc990c921df7eb4c20355a5

### CODE REVIEW

File: vllm_ascend/quantization/methods/fp8.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+from vllm.config import get_current_vllm_config
+
+from .base import QuantType
+from .w8a8_mxfp8 import AscendW8A8MXFP8DynamicLinearMethod
+from .w4a8_mxfp4 import AscendW4A8MXFPDynamicFusedMoEMethod
+from .registry import register_scheme
+
+
+@register_scheme("FP8", "ds_linear")
+class AscendW8A8MXFP8DSDynamicLinearMethod(AscendW8A8MXFP8DynamicLinearMethod):
```
Comment (lcfenglinwan) [inline vllm_ascend/quantization/methods/fp8.py:31]: 添加类的说明，针对fp8原生适配

File: vllm_ascend/quantization/method_adapters.py [hunk 2]
```diff
@@ -110,6 +110,11 @@ def create_weights(
             set_weight_attrs(param, {"output_dim": 0})
             layer.register_parameter(pergroup_name, param)
             set_weight_attrs(param, extra_weight_attrs)
+            set_weight_attrs(param, {
```
Comment (lcfenglinwan) [inline vllm_ascend/quantization/method_adapters.py]: check是否会影响其他模型

File: vllm_ascend/quantization/fp8_config.py
Comment (lcfenglinwan) [inline vllm_ascend/quantization/fp8_config.py:1]: check vllm 上游 fp8实现，分析实现差异

File: vllm_ascend/quantization/method_adapters.py (hunk 2 above)
Comment (lcfenglinwan) [reply vllm_ascend/quantization/method_adapters.py]: 已修改，只针对有pack_factor相关属性的场景下配置相关属性

File: vllm_ascend/quantization/methods/fp8.py (hunk 1 above)
Comment (lcfenglinwan) [reply vllm_ascend/quantization/methods/fp8.py:31]: 已添加

**Review by linfeng-yuan — APPROVED** [review]
LGTM

File: vllm_ascend/models/deepseek_v4.py [hunk 3]
```diff
@@ -497,19 +500,22 @@ def __init__(
...
             self.coff * self.head_dim,
             bias=False,
-            quant_config=quant_config,
+            quant_config=None if get_ascend_device_type() in {AscendDeviceType.A5} else quant_config,
             prefix=f"{prefix}.wgate",
             return_bias=False,
         )
-        self.norm = RMSNorm(self.head_dim, config.rms_norm_eps)
+
+        # A5 compressor算子的norm weight入参数据类型为float
```
Comment (MengqingCao) [inline vllm_ascend/models/deepseek_v4.py]: plz use English

File: vllm_ascend/ops/activation.py [hunk 4]
```diff
@@ -40,6 +40,21 @@ def forward_oot(self, x: torch.Tensor) -> torch.Tensor:
         return out
 
 
+class AscendSiluAndMulWithClamp(SiluAndMulWithClamp):
+    def forward_oot(self, x: torch.Tensor) -> torch.Tensor:
+        import torch_npu
```
Comment (MengqingCao) [inline vllm_ascend/ops/activation.py]: don't do this lazy init in a forward of operator

**Review by MengqingCao — APPROVED** [review]

File: vllm_ascend/models/deepseek_v4.py (hunk 3 above)
Comment (lcfenglinwan) [reply vllm_ascend/models/deepseek_v4.py]: done

File: vllm_ascend/ops/activation.py (hunk 4 above)
Comment (lcfenglinwan) [reply vllm_ascend/ops/activation.py]: done

**Review by linfeng-yuan — APPROVED** [review]
LGTM. And CI failed with known issues (e.g., test_shared_expert_dp accuracy, OOM, non-EP FC assertion). Let's merge it first
